src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -0,0 +1,350 @@
+package io.iohk.ethereum.blockchain.sync
+
+import scala.concurrent.duration._
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.ExecutionContext.Implicits.global
+
+import akka.actor._
+import akka.util.ByteString
+import io.iohk.ethereum.domain.BlockHeader
+import io.iohk.ethereum.network.PeerActor
+import io.iohk.ethereum.network.PeerActor.Status.{Handshaked, Chain}
+import io.iohk.ethereum.network.p2p.messages.PV62.{GetBlockHeaders, BlockHeaders}
+import io.iohk.ethereum.utils.Config.FastSync._
+
+trait FastSync {
+  selfSyncController: SyncController =>
+
+  import FastSync._
+  import SyncController._
+
+  def startFastSync(): Unit = {
+    log.info("Starting fast sync")
+    fastSyncStateStorage.getSyncState() match {
+      case Some(syncState) => startFastSync(syncState)
+      case None => startFastSyncFromScratch()
+    }
+  }
+
+  private def startingFastSync: Receive = handlePeerUpdates orElse {
+    case RetryStart => startFastSync()
+  }
+
+  private def startFastSync(syncState: SyncState): Unit = {
+    scheduler.schedule(0.seconds, printStatusInterval, self, PrintStatus)
+    context become new SyncingHandler(syncState).receive
+    self ! ProcessSyncing
+  }
+
+  private def startFastSyncFromScratch(): Unit = {
+    val peersUsedToChooseTarget = peersToDownloadFrom.filter(_._2.chain == Chain.ETC)
+
+    if (peersUsedToChooseTarget.size >= minPeersToChooseTargetBlock) {
+      peersUsedToChooseTarget.foreach { case (peer, Handshaked(status, _, _)) =>
+        peer ! PeerActor.Subscribe(Set(BlockHeaders.code))
+        peer ! PeerActor.SendMessage(GetBlockHeaders(Right(status.bestHash), 1, 0, reverse = false))
+      }
+      log.info("Asking {} peers for block headers", peersUsedToChooseTarget.size)
+      val timeout = scheduler.scheduleOnce(peerResponseTimeout, self, BlockHeadersTimeout)
+      context become waitingForBlockHeaders(peersUsedToChooseTarget.keySet, Map.empty, timeout)
+    } else {
+      log.warning("Cannot start fast sync, not enough peers to download from. Scheduling retry in {}", startRetryInterval)
+      scheduleStartRetry(startRetryInterval)
```

Comment:
Should we do here `context become startingFastSync`? I'm not sure whether we can handle the `RetryStart` message if that's not the case, I wasn't able to start our node without this line.